// @keystonejs/field-content/dist/Field-af2b19a5.esm.js
import '@babel/runtime/helpers/esm/objectWithoutProperties';
import { Block } from 'slate';
import _extends from '@babel/runtime/helpers/esm/extends';
import { jsx } from '@emotion/core';
import { g as getSelectionReference } from './utils-1c16766b.esm.js';
import { PlusIcon, CircleSlashIcon } from '@arch-ui/icons';
import { colors, gridSize } from '@arch-ui/theme';
import '@arch-ui/color-utils';
import '@arch-ui/tooltip';
import '@arch-ui/typography';
import { T as ToolbarButton } from './toolbar-components-93e85906.esm.js';
import React__default, { useRef, useCallback, useLayoutEffect, useState, Fragment, forwardRef, useMemo } from 'react';
import { type } from './paragraph-531e7b46.esm.js';
import { Popper } from 'react-popper';
import { a as useWindowSize, b as useScrollListener, u as useStateWithEqualityCheck } from './index-2a33fff9.esm.js';
import { Editor } from 'slate-react';
import { createPortal } from 'react-dom';
import isHotkey from 'is-hotkey';
import { getVisibleSelectionRect } from 'get-selection-range';
import { useMeasure } from '@arch-ui/hooks';
import applyRef from 'apply-ref';
import { FieldContainer, FieldLabel, FieldInput } from '@arch-ui/fields';
import { inputStyles } from '@arch-ui/input';

function markPlugin(type, options) {
  return {
    onKeyDown(event, editor, next) {
      // If it doesn't match our `key`, let other plugins handle it.
      if (!options.test(event)) return next(); // Prevent the default characters from being inserted.

      event.preventDefault(); // Toggle the mark `type`.

      editor.toggleMark(type);
    },

    renderMark: (props, editor, next) => {
      if (props.mark.type === type) {
        return options.render(props);
      }

      return next();
    }
  };
}

let marks = {
  bold: {
    test: isHotkey('mod+b'),
    label: 'Bold',
    icon: props => {
      return React__default.createElement("strong", _extends({}, props, {
        "aria-hidden": true
      }), "B");
    },
    render: props => React__default.createElement("strong", props.attributes, props.children)
  },
  italic: {
    test: isHotkey('mod+i'),
    label: 'Italic',
    icon: props => {
      return React__default.createElement("em", _extends({}, props, {
        "aria-hidden": true
      }), "I");
    },
    render: props => React__default.createElement("em", props.attributes, props.children)
  },
  strikethrough: {
    test: isHotkey('mod+~'),
    label: 'Strikethrough',
    icon: props => {
      return React__default.createElement("s", _extends({}, props, {
        "aria-hidden": true
      }), "S");
    },
    render: props => React__default.createElement("s", props.attributes, props.children)
  },
  underline: {
    test: isHotkey('mod+u'),
    label: 'Underline',
    icon: props => {
      return React__default.createElement("u", _extends({}, props, {
        "aria-hidden": true
      }), "U");
    },
    render: props => React__default.createElement("u", props.attributes, props.children)
  }
};
let markTypes = Object.keys(marks);
let plugins = Object.entries(marks).map(_ref => {
  let [type, options] = _ref;
  return markPlugin(type, options);
});
/** @jsx jsx */

let AddBlock = _ref => {
  let {
    editorState,
    editor,
    blocks
  } = _ref;
  let windowSize = useWindowSize();
  let openCloseRef = useRef(null);
  let containerRef = useRef(null);
  let focusBlock = editorState.focusBlock;
  let layout = useCallback(() => {
    let openCloseEle = openCloseRef.current;
    let containerEle = containerRef.current;
    const rect = getVisibleSelectionRect();

    if (!rect || rect.width !== 0 || focusBlock === null || focusBlock.text !== '' || focusBlock.type !== type) {
      setIsOpen(false);
      openCloseEle.style.top = '';
      openCloseEle.style.left = '';
      containerEle.style.top = '';
      containerEle.style.left = '';
      return;
    }

    const top = rect.top + window.scrollY - openCloseEle.offsetHeight / 2 + rect.height / 2; // eslint-disable-line

    openCloseEle.style.top = "".concat(top, "px");
    containerEle.style.top = "".concat(top, "px");
    const containerEleLeft = rect.left + window.scrollX;
    containerEle.style.left = "".concat(containerEleLeft, "px");
    const left = containerEleLeft - openCloseEle.offsetWidth;
    openCloseEle.style.left = "".concat(left, "px");
  }, [focusBlock, windowSize]);
  useLayoutEffect(layout);
  useScrollListener(layout);
  let [isOpen, setIsOpen] = useState(false);
  return createPortal(jsx(Fragment, null, jsx("div", {
    ref: openCloseRef,
    css: {
      position: 'absolute',
      top: -10000,
      left: -10000
    }
  }, jsx("button", {
    type: "button",
    css: {
      borderRadius: '100%',
      border: '1px black solid',
      width: 30,
      height: 30,
      marginRight: 4
    },
    onClick: () => {
      setIsOpen(x => !x);
    }
  }, jsx(PlusIcon, {
    css: {
      transition: '50ms transform',
      transform: isOpen ? 'rotateZ(45deg)' : 'rotateZ(0deg)'
    },
    title: isOpen ? 'Close' : 'Open'
  }))), jsx("div", {
    css: {
      position: 'absolute',
      top: -10000,
      left: -10000
    },
    ref: containerRef
  }, isOpen && jsx("div", null, Object.keys(blocks).map(key => {
    let {
      Sidebar
    } = blocks[key];

    if (!blocks[key].withChrome || Sidebar === undefined) {
      return null;
    }

    return jsx(Sidebar, {
      key: key,
      editor: editor,
      blocks: blocks
    });
  })))), document.body);
};
/** @jsx jsx */


let stopPropagation = e => {
  e.stopPropagation();
};

function InnerToolbar(_ref) {
  let {
    blocks,
    editor,
    editorState
  } = _ref;
  return jsx("div", {
    css: {
      display: 'flex'
    }
  }, Object.keys(blocks).map(x => blocks[x].withChrome && blocks[x].Toolbar).filter(x => x).reduce((children, Toolbar) => {
    return jsx(Toolbar, {
      editor: editor,
      editorState: editorState
    }, children);
  }, jsx(Fragment, null, Object.keys(marks).map(name => {
    let Icon = marks[name].icon;
    return jsx(ToolbarButton, {
      label: marks[name].label,
      icon: jsx(Icon, null),
      isActive: editorState.activeMarks.some(mark => mark.type === name),
      onClick: () => {
        editor.toggleMark(name);
        editor.focus();
      },
      key: name
    });
  }), jsx(ToolbarButton, {
    label: "Remove Formatting",
    icon: jsx(CircleSlashIcon, null),
    onClick: () => {
      markTypes.forEach(mark => {
        editor.removeMark(mark);
      });
      editor.focus();
    }
  }), Object.keys(blocks).map(type => {
    let ToolbarElement = blocks[type].ToolbarElement;

    if (!blocks[type].withChrome || ToolbarElement === undefined) {
      return null;
    }

    return jsx(ToolbarElement, {
      key: type,
      editor: editor,
      editorState: editorState
    });
  }))));
}

const PopperRender = forwardRef((_ref2, _ref3) => {
  let {
    scheduleUpdate,
    editorState,
    style,
    children
  } = _ref2;
  let {
    fragment
  } = editorState;
  let shouldShowToolbar = fragment.text !== '';
  let containerRef = useRef(null);
  let snapshot = useMeasure(containerRef);
  useLayoutEffect(() => {
    if (shouldShowToolbar) {
      scheduleUpdate();
    }
  }, [scheduleUpdate, editorState, snapshot, shouldShowToolbar]);
  return createPortal(jsx("div", {
    onMouseDown: stopPropagation,
    ref: node => {
      applyRef(_ref3, node);
      applyRef(containerRef, node);
    },
    style: style,
    css: {
      // this isn't as nice of a transition as i'd like since the time is fixed
      // i think it would better if it was physics based but that would probably
      // be a lot of work for little gain
      // maybe base the transition time on the previous value?
      transition: 'transform 100ms, opacity 100ms'
    }
  }, jsx("div", {
    css: {
      backgroundColor: colors.N90,
      padding: 8,
      borderRadius: 6,
      margin: gridSize,
      display: shouldShowToolbar ? 'flex' : 'none'
    }
  }, shouldShowToolbar && children)), document.body);
});

var Toolbar = _ref4 => {
  let {
    editorState,
    blocks,
    editor
  } = _ref4; // this element is created here so that when the popper rerenders
  // the inner toolbar won't have to update

  let children = jsx(InnerToolbar, {
    blocks: blocks,
    editor: editor,
    editorState: editorState
  });
  return jsx(Popper, {
    placement: "top",
    referenceElement: // the reason we do this rather than having the selection reference
    // be constant is because the selection reference
    // has some internal state and it shouldn't persist between different
    // editor references
    useMemo(getSelectionReference, [])
  }, _ref5 => {
    let {
      style,
      ref,
      scheduleUpdate
    } = _ref5;
    return jsx(PopperRender, {
      scheduleUpdate,
      editorState,
      style,
      blocks,
      editor,
      ref,
      children
    });
  });
};

function getSchema(blocks) {
  const schema = {
    document: {
      last: {
        type: type
      },
      normalize: (editor, _ref) => {
        let {
          code,
          node
        } = _ref;

        switch (code) {
          case 'last_child_type_invalid':
            {
              const paragraph = Block.create(type);
              return editor.insertNodeByKey(node.key, node.nodes.size, paragraph);
            }
        }
      }
    },
    blocks: {}
  };
  Object.keys(blocks).forEach(type => {
    if (typeof blocks[type].getSchema === 'function') {
      schema.blocks[type] = blocks[type].getSchema({
        blocks
      });
    }
  });
  return schema;
}

function Stories(_ref2) {
  let {
    value: editorState,
    onChange: _onChange,
    blocks,
    className
  } = _ref2;
  let schema = useMemo(() => {
    return getSchema(blocks);
  }, [blocks]);
  let plugins$1 = useMemo(() => {
    const renderNode = props => {
      let block = blocks[props.node.type];

      if (block) {
        return jsx(block.Node, _extends({}, props, {
          blocks: blocks
        }));
      }

      return null;
    };

    return Object.values(blocks).reduce((combinedPlugins, block) => {
      if (typeof block.getPlugins !== 'function') {
        return combinedPlugins;
      }

      return [...combinedPlugins, ...block.getPlugins({
        blocks
      })];
    }, [...plugins, {
      renderBlock: renderNode,
      renderInline: renderNode
    }]);
  }, [blocks]);
  let [editor, setEditor] = useStateWithEqualityCheck(null);
  return jsx("div", {
    className: className
  }, jsx(Editor, {
    schema: schema,
    ref: setEditor,
    plugins: plugins$1,
    value: editorState,
    onChange: _ref3 => {
      let {
        value
      } = _ref3;

      _onChange(value);
    }
  }), jsx(AddBlock, {
    editor: editor,
    editorState: editorState,
    blocks: blocks
  }), jsx(Toolbar, {
    editorState,
    editor,
    blocks
  }));
}
/** @jsx jsx */


let ContentField = _ref => {
  let {
    field,
    value,
    onChange,
    autoFocus,
    errors
  } = _ref;
  const htmlID = "ks-content-editor-".concat(field.path);
  return jsx(FieldContainer, null, jsx(FieldLabel, {
    htmlFor: htmlID,
    field: field,
    errors: errors
  }), jsx(FieldInput, null, Object.values(field.getBlocks()).filter(_ref2 => {
    let {
      Provider,
      options
    } = _ref2;
    return Provider && options;
  }).reduce((children, _ref3, index) => {
    let {
      Provider,
      options
    } = _ref3;
    return (// Using index within key is ok here as the blocks never change
      // across renders
      jsx(Provider, {
        value: options,
        key: "".concat(htmlID, "-provider-").concat(index)
      }, children)
    );
  }, jsx(Stories, {
    key: htmlID,
    blocks: field.getBlocks(),
    value: value,
    onChange: onChange,
    autoFocus: autoFocus,
    id: htmlID,
    css: inputStyles({
      isMultiline: true
    })
  }))));
};

export default ContentField;
