// @arch-ui/select/dist/select.cjs.prod.js
"use strict";

function _interopDefault(ex) {
  return ex && "object" == typeof ex && "default" in ex ? ex.default : ex;
}

Object.defineProperty(exports, "__esModule", {
  value: !0
});

var _extends = _interopDefault(require("@babel/runtime/helpers/extends")), _objectWithoutProperties = _interopDefault(require("@babel/runtime/helpers/objectWithoutProperties")), _objectSpread = _interopDefault(require("@babel/runtime/helpers/objectSpread")), React = require("react"), BaseSelect = _interopDefault(require("react-select")), AsyncCreatableSelect = _interopDefault(require("react-select/async-creatable")), AsyncSelect = _interopDefault(require("react-select/async")), CreatableSelect = _interopDefault(require("react-select/creatable")), theme = require("@arch-ui/theme");

const indicatorStyles = (provided, _ref) => {
  let isDisabled = _ref.isDisabled, isFocused = _ref.isFocused, styles = {
    color: theme.colors.N20,
    ":hover": isDisabled || isFocused ? null : {
      color: theme.colors.N40
    }
  };
  return isDisabled && (styles = {
    color: theme.colors.N10
  }), isFocused && (styles = {
    color: theme.colors.N60,
    ":hover": {
      color: theme.colors.N80
    }
  }), _objectSpread({}, provided, styles);
}, selectStyles = {
  control: (provided, _ref2) => {
    const focusStyles = _ref2.isFocused ? {
      borderColor: theme.colors.primary,
      boxShadow: "inset 0 1px 1px rgba(0, 0, 0, 0.075),\n      0 0 0 3px ".concat(theme.colors.B.A20),
      outline: 0
    } : null;
    return _objectSpread({}, provided, {
      backgroundColor: "white",
      borderColor: theme.colors.N20,
      fontSize: "1rem",
      minHeight: 35,
      minWidth: "200px",
      ":hover": {
        borderColor: theme.colors.N30
      }
    }, focusStyles);
  },
  clearIndicator: indicatorStyles,
  dropdownIndicator: indicatorStyles,
  menu: provided => _objectSpread({}, provided, {
    fontSize: "0.9em"
  }),
  option: (provided, _ref3) => {
    let isDisabled = _ref3.isDisabled, isFocused = _ref3.isFocused, isSelected = _ref3.isSelected, bg = "inherit";
    isFocused && (bg = theme.colors.B.L90), isSelected && (bg = theme.colors.primary);
    let txt = "inherit";
    isFocused && (txt = theme.colors.primary), isSelected && (txt = "white"), isDisabled && (txt = theme.colors.N40);
    const cssPseudoActive = isSelected || isDisabled ? {} : {
      backgroundColor: theme.colors.B.L80,
      color: theme.colors.B.D20
    };
    return _objectSpread({}, provided, {
      fontSize: "1rem",
      backgroundColor: bg,
      color: txt,
      ":active": cssPseudoActive
    });
  },
  menuPortal: provided => _objectSpread({}, provided, {
    zIndex: 3
  })
}, getSelectVariant = _ref4 => {
  let isAsync = _ref4.isAsync, isCreatable = _ref4.isCreatable;
  return isAsync && isCreatable ? AsyncCreatableSelect : isAsync ? AsyncSelect : isCreatable ? CreatableSelect : BaseSelect;
}, Select = _ref5 => {
  let isAsync = _ref5.isAsync, isCreatable = _ref5.isCreatable, innerRef = _ref5.innerRef, styles = _ref5.styles, props = _objectWithoutProperties(_ref5, [ "isAsync", "isCreatable", "innerRef", "styles" ]);
  const ReactSelect = getSelectVariant({
    isAsync: isAsync,
    isCreatable: isCreatable
  });
  return React.createElement(ReactSelect, _extends({
    ref: innerRef,
    styles: React.useMemo(() => _objectSpread({}, selectStyles, styles), [ styles ])
  }, props));
};

exports.default = Select;
